### DISCUSSION TITLE
Troubleshooting linker error when compiling AI-deck wifi-img-streamer example with JPEG_ENCODING using the gap_sdk

### DISCUSSION THREAD

**Discussion by gelanchez** [bitcraze/discussions#1323] [Q&A] [answered]
Hi,

When trying to compile the **wifi-img-streamer** example with the JPEG_ENCODING enabled and the gap_sdk locally installed, I'm getting the following linker error:

```
/home/jose/projects/aideck-gap8-examples/examples/other/wifi-img-streamer/BUILD/GAP8_V2/GCC_RISCV_FREERTOS//home/jose/gap_sdk/libs/gap_lib/jpeg/cluster.o: In function `pop_ready_block':
/home/jose/gap_sdk/libs/gap_lib/jpeg/cluster.c:138: undefined reference to `hal_cl_eu_event_mask_wait_clear'
/home/jose/gap_sdk/libs/gap_lib/jpeg/cluster.c:138: undefined reference to `hal_cl_eu_event_mask_wait_clear'
collect2: error: ld returned 1 exit status
make: *** [/home/jose/gap_sdk/rtos/freeRTOS/vendors/gwt/rules/freeRTOS_rules.mk:419: /home/jose/projects/aideck-gap8-examples/examples/other/wifi-img-streamer/BUILD/GAP8_V2/GCC_RISCV_FREERTOS/wifi-img-streamer] Error 1
```

However the compilation works well with the RAW_ENCODING and the gap_sdk, and also with both encoding options if I use the docker image. Do you have any ideas why this happens? Are there any environment variables used in the docker image I'm missing? Any additional flags I need to add? Did I miss anything when I compiled the sdk?

Many thanks!

**Comment by gemenerik** [comment]
What version of the gap_sdk did you install? Or are you on master? 
Do you source the aideck config?
Did you build all parts of the sdk?

The gap_sdk in the docker has some changes, most of them related to the offset for our bootloader. Other changes (hyperram and hyperflash includes and defines) also seem unrelated to the error you are getting, but you could give it a shot. You can git diff inside the /gap_sdk folder in the docker to see all changes.

**Comment by gelanchez** [comment]
Many thanks, @gemenerik,

I'm on the gap_sdk master, so latest version, I did source the aideck (_configs/ai_deck.sh_) and I think I've built all parts of the SDK by successfully running a `make clean sdk` there.

A `git diff` in the gap_sdk folder of the docker image only mentions three changes changes in _flash_partition.c_ and a small one  on a .PHONY in _freeRTOS_rules.mk_. I don't think these are related to the linker error (not sure if you meant those). A `git diff` in my local gap_sdk only shows a couple I made to make it work with the JTAG cable I'm using.

**Reply by gemenerik** [reply]
Can you try manually removing the build folder of the example? So `/home/jose/projects/aideck-gap8-examples/examples/other/wifi-img-streamer/BUILD` in your case. And then try to compile it again.

**Reply by gelanchez** [reply]
Thanks, Rik. I get the same error after deleting the BUILD folder. It goes away if I remove the line:
`jpeg_encoder_process(&jpeg_encoder, &buffer, &jpeg_data, &jpegSize);`
I'm not sure how they can be related.

**Reply by gelanchez** [reply]
Hi, @gemenerik,

Which gap_sdk version is it used in the docker?
The file libs/gap_lib/jpeg/cluster.c in the latest SDK (4.12)  starts with a:
```
#ifdef PMSIS_DRIVERS
    #define RT_USER_EVENT (CL_USER_EVENT)
    #define eu_evt_trig_from_id(x,y) (hal_eu_cluster_evt_trig_set(x,y))
#if defined(__GAP9__)
    #define eu_evt_maskWaitAndClr(x) (hal_cl_eu_event_mask_wait_and_clear(x))
#else
    #define eu_evt_maskWaitAndClr(x) (hal_cl_eu_event_mask_wait_clear(x))
#endif
```

However the same file in the docker is:
```
#ifdef PMSIS_DRIVERS
    #define RT_USER_EVENT (CL_USER_EVENT)
    #define eu_evt_trig_from_id(x,y) (hal_eu_cluster_evt_trig_set(x,y))
    #define eu_evt_maskWaitAndClr(x) (hal_cl_eu_evt_mask_wait_clear(x))
#endif
```
Since the error is that the reference to _hal_cl_eu_event_mask_wait_clear(x)_ is not defined, could it be a bug in the SDK?

**Reply by knmcguire** [reply]
Rick is away for now, but currently we are on gapsdk v4.22.0 in docker as of a few weeks.

Could you pulled the lastest version of the docker image and try again?

**Reply by miferco97** [reply]
Hello all

> Could you pulled the latest version of the docker image and try again?

I have done it from fresh install today and have the same issue :(

**Reply by gelanchez** [reply]
Thanks for confirming, @miferco97. Then I'll keep the old version active until the issue is fixed, as that one compiles.

**Reply by knmcguire** [reply]
Alright, Rik should be back monday again so I'll let him know about this.

**Reply by gemenerik** [reply]
> > Could you pulled the latest version of the docker image and try again?
> 
> I have done it from fresh install today and have the same issue :(

I just verified that the streamer demo compiles and runs. Can you confirm you are on the latest docker image? Since this example does not require the autotiler you can run from `bitcraze/aideck:latest` instead of your `aideck-with-autotiler` (make sure you update this container based on your latest pulled `bitcraze/aideck` if you plan to use autotiler later).

```
docker run --rm -v ${PWD}:/module bitcraze/aideck:latest tools/build/make-example examples/other/wifi-img-streamer image
```

If instead you still want to use your sdk from source: It may be a bit confusing, the gap_sdk repo has tags but no "GitHub releases" for the v4.22.0 release. Try pulling that tag or even the latest master instead and let me know if the streamer demo compiles.

I recommended deleting the BUILD folder because make clean doesn't actually clean the entire build. I have had problems before with upgrading my sdk.

**Reply by gelanchez** [reply]
Hi, @gemenerik,

Thanks. The gap_sdk I'm using must be the latest version, as I got it with a `git clone` from the repository. That one doesn't work.

I'll try the latest docker image, but since the one I have now compiles, before updating to one that doesn't, can you confirm the example works on your side with the  JPEG_ENCODING enabled, as that doesn't come by default and it's the bit giving problems? You might have tried without that, which works fine.

**Reply by gemenerik** [reply, marked as answer]
Sorry, you are right, with JPEG encoding enabled I get the same error. I fixed it in docker image `bitcraze/aideck:4.22.0.1` and opened [a pr](https://github.com/GreenWaves-Technologies/gap_sdk/pull/417) on the `gap_sdk` repository.

You can easily switch between versions of the docker image (bitcraze/aideck:tag with tags found [here](https://hub.docker.com/r/bitcraze/aideck/tags)) and between docker and your local installed sdk so no need to worry about trying things.

**Reply by gelanchez** [reply]
Thanks a lot, @gemenerik! Not sure how long it will take the PR to be approved, but now we have a solution. I'll set the discussion as resolved.
